src/core/dom/layout/viewport-observer.js

```diff
@@ -80,8 +76,11 @@ export function unobserveWithSharedInOb(element) {
  * @param {!Array<!IntersectionObserverEntry>} entries
  */
 function ioCallback(entries) {
-  for (let i = 0; i < entries.length; i++) {
-    const {isIntersecting, target} = entries[i];
-    viewportCallbacks.get(target)?.(isIntersecting);
-  }
+  entries.forEach((entry) => {
+    const {target} = entry;
+    const callbacks = viewportCallbacks.get(target) ?? [];
+    callbacks.forEach((callback) => {
```

Comment:
Always prefer for loops over `forEach`. This creates `n * m` closures where none is needed.